## RNN

## binary with my model

In [68]:
import pandas as pd
import numpy as np
import copy
import re
import torch
from sklearn.model_selection import train_test_split
import random

In [83]:
import time

In [2]:
import pandas as pd
test = pd.read_csv("amazon_reviews_us_Kitchen_v1_00.tsv", sep = '\t',error_bad_lines = False)
test['label'] = -1
a1 = test.loc[test['star_rating']==1].sample(50000)
a2 = test.loc[test['star_rating']==2].sample(50000)
a3 = test.loc[test['star_rating']==3].sample(50000)
a4 = test.loc[test['star_rating']==4].sample(50000)
a5 = test.loc[test['star_rating']==5].sample(50000)

b'Skipping line 16148: expected 15 fields, saw 22\nSkipping line 20100: expected 15 fields, saw 22\nSkipping line 45178: expected 15 fields, saw 22\nSkipping line 48700: expected 15 fields, saw 22\nSkipping line 63331: expected 15 fields, saw 22\n'
b'Skipping line 86053: expected 15 fields, saw 22\nSkipping line 88858: expected 15 fields, saw 22\nSkipping line 115017: expected 15 fields, saw 22\n'
b'Skipping line 137366: expected 15 fields, saw 22\nSkipping line 139110: expected 15 fields, saw 22\nSkipping line 165540: expected 15 fields, saw 22\nSkipping line 171813: expected 15 fields, saw 22\n'
b'Skipping line 203723: expected 15 fields, saw 22\nSkipping line 209366: expected 15 fields, saw 22\nSkipping line 211310: expected 15 fields, saw 22\nSkipping line 246351: expected 15 fields, saw 22\nSkipping line 252364: expected 15 fields, saw 22\n'
b'Skipping line 267003: expected 15 fields, saw 22\nSkipping line 268957: expected 15 fields, saw 22\nSkipping line 303336: expected 15 field

b'Skipping line 1968846: expected 15 fields, saw 22\nSkipping line 1999941: expected 15 fields, saw 22\nSkipping line 2001492: expected 15 fields, saw 22\nSkipping line 2011204: expected 15 fields, saw 22\nSkipping line 2025295: expected 15 fields, saw 22\n'
b'Skipping line 2041266: expected 15 fields, saw 22\nSkipping line 2073314: expected 15 fields, saw 22\nSkipping line 2080133: expected 15 fields, saw 22\nSkipping line 2088521: expected 15 fields, saw 22\n'
b'Skipping line 2103490: expected 15 fields, saw 22\nSkipping line 2115278: expected 15 fields, saw 22\nSkipping line 2153174: expected 15 fields, saw 22\nSkipping line 2161731: expected 15 fields, saw 22\n'
b'Skipping line 2165250: expected 15 fields, saw 22\nSkipping line 2175132: expected 15 fields, saw 22\nSkipping line 2206817: expected 15 fields, saw 22\nSkipping line 2215848: expected 15 fields, saw 22\nSkipping line 2223811: expected 15 fields, saw 22\n'
b'Skipping line 2257265: expected 15 fields, saw 22\nSkipping line

In [3]:
new_test = pd.concat([a1,a2,a3,a4,a5])

In [4]:
new_test['lable'] = 1

In [5]:
new_test.label[test.star_rating>3] = 1
new_test.label[test.star_rating<3] = 2
new_test.label[test.star_rating==3] = 3
new_test = new_test[['label','review_body']] 

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so w

In [6]:
import gensim
model_1 = gensim.models.Word2Vec.load('save')

In [7]:
#data clearing
new_test['review_body'] = new_test['review_body'].str.lower()

def tag(x):
    return re.sub('<.*?>','',str(x))
new_test['review_body'] = new_test['review_body'].apply(lambda x:tag(x))

def url(x):
    return re.sub('(https?|ftp|file)://[-A-Za-z0-9+&@#/%?=~_|!:,.;]+[-A-Za-z0-9+&@#/%=~_|]','',str(x))

new_test['review_body'] = new_test['review_body'].apply(lambda x:url(x))

import contractions
new_test['review_body'] = new_test['review_body'].apply(lambda x:contractions.fix(x))

def non_alphabetical(x):
    return re.sub('[^a-zA-Z\s]','',str(x))
new_test['review_body'] = new_test['review_body'].apply(lambda x:non_alphabetical(x))
def extra_space(x):
    return re.sub( ' +',' ',str(x))
new_test['review_body'] = new_test['review_body'].apply(lambda x:extra_space(x))

In [8]:
#Pre-processing

import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')
stop_words_set = set(stopwords.words('english'))
from nltk import word_tokenize, pos_tag

def stop_words(x):
    word_tokens = word_tokenize(x)
    temp = []
    for i in word_tokens:
        if i not in stop_words_set:
            temp.append(i)
    return temp
new_test['review_body'] = new_test['review_body'].apply(lambda x:stop_words(x))

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/chenlin/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [10]:
new_test['y'] = -1

In [11]:
new_test['y'][new_test['label'] == 1] =0

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [12]:
new_test['y'][new_test['label'] == 2] =1

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [13]:
new_test['y'][new_test['label'] == 3] =2

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [14]:
binary = new_test.loc[new_test['y']!=2]

In [15]:
binary

label                                        review_body  y
1674978      2                           [wrong, sizehad, return]  1
3055998      2  [bought, replace, old, mandolins, slicer, cut,...  1
4165773      2  [looking, forward, getting, handy, new, kitche...  1
2168375      2  [item, cheap, could, made, paper, inside, croo...  1
2627653      2  [got, year, old, boyfriend, awesome, girlfrien...  1
...        ...                                                ... ..
4171764      1  [made, amazing, popcorn, hot, air, corn, poppe...  0
4272127      1  [bought, bagel, slicer, kids, would, cut, slic...  0
1302570      1                                      [great, deal]  0
2914686      1  [original, quality, quart, bowl, kitchenaid, r...  0
3063914      1  [straws, great, little, kids, keep, chewing, c...  0

[200000 rows x 3 columns]

In [210]:
def mymodel_reviewtensor(x):
    max_review = 50
    ### create a matrix 
    review = torch.zeros(max_review,1,300)
    pad = torch.zeros(300)
    index = 0
    for i in x:
        if i in model_1.wv and index < 50:
            vector = model_1.wv [i]
            review[index][0] = torch.from_numpy(vector)
            index += 1
        else:
            continue
    if index < 50:
        review[index][0] = pad
    review = torch.nan_to_num(review)
    return review 

In [213]:
X = binary['review_body']
Y = binary['y']

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 19, test_size = 0.2)

x_train = np.array(x_train.tolist())

y_train = np.array(y_train.tolist())

x_train = np.nan_to_num(x_train)

x_test = np.array(x_test.tolist())

y_test = np.array(y_test.tolist())

x_test = np.nan_to_num(x_test)

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:8: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:14: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  


In [214]:
from torch.utils.data import DataLoader, Dataset
from torch.utils.data import TensorDataset, DataLoader

In [215]:
import torch.nn as nn

class RNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(RNN, self).__init__()

        self.hidden_size = hidden_size

        self.i2h = nn.Linear(input_size + hidden_size, hidden_size)
        self.i2o = nn.Linear(input_size + hidden_size, output_size)
        self.softmax = nn.LogSoftmax(dim=1)

    def forward(self, input, hidden):
        combined = torch.cat((input, hidden), 1)
        hidden = self.i2h(combined)
        output = self.i2o(combined)
        output = self.softmax(output)
        return output, hidden

    def initHidden(self):
        return torch.zeros(1, self.hidden_size)

rnn = RNN(300, 50, 2)

In [222]:
learning_rate = 0.005 # If you set this too high, it might explode. If too low, it might not learn

def train(category_tensor, line_tensor):
    hidden = rnn.initHidden()

    rnn.zero_grad()

    for i in range(line_tensor.size()[0]):
        output, hidden = rnn(line_tensor[i], hidden)

    loss = criterion(output, category_tensor)
    loss.backward()
    torch.nn.utils.clip_grad_norm_(rnn.parameters(),0.5)
    # Add parameters' gradients to their values, multiplied by learning rate
    for p in rnn.parameters():
        p.data.add_(p.grad.data, alpha=-learning_rate)

    return output, loss.item()

In [216]:
input = mymodel_reviewtensor(binary['review_body'].iloc[4])
hidden = torch.zeros(1, 50)

In [217]:
def output_predict_y(output):
    top_n, top_i = output.topk(1)
    y = top_i[0].item()
    return y

In [218]:
criterion = nn.NLLLoss()

In [219]:
main_data = list(zip(x_train,y_train))

In [220]:
# Keep track of losses for plotting
current_loss = 0

def timeSince(since):
    now = time.time()
    s = now - since
    m = math.floor(s / 60)
    s -= m * 60
    return '%dm %ds' % (m, s)

start = time.time()
ii=0
correct = 0
for x_i,y_i in main_data:
    line_tensor = mymodel_reviewtensor(x_i)
    y_i = torch.tensor([y_i],dtype =torch.long)
    output, loss = train(y_i, line_tensor)
    current_loss += loss
    ii += 1
    if ii %2500 == 0 :
        print(ii,"current avg lose is ",current_loss/ii,'time is ',time.time() - start)

2500 current avg lose is  0.6935066653609275 time is  18.736644983291626
5000 current avg lose is  0.6938000483036041 time is  39.694331884384155
7500 current avg lose is  0.6940343046387036 time is  59.080775022506714
10000 current avg lose is  0.6943283890724182 time is  75.36496996879578
12500 current avg lose is  0.6985913027977944 time is  93.71864986419678
15000 current avg lose is  0.706857127382358 time is  113.98635816574097
17500 current avg lose is  0.7157444509387016 time is  133.35544085502625
20000 current avg lose is  0.719268296802044 time is  151.9071388244629
22500 current avg lose is  0.7231831370088789 time is  167.7384889125824
25000 current avg lose is  0.7257247288429737 time is  185.27695298194885
27500 current avg lose is  0.7280170371738347 time is  204.13538002967834
30000 current avg lose is  0.7308424241900444 time is  220.9835648536682
32500 current avg lose is  0.7335039394947199 time is  237.65777802467346
35000 current avg lose is  0.736566359599999 tim

In [223]:

n_label = 2
confusion = torch.zeros(n_label, n_label)

def evaluate(line_tensor):
    hidden = rnn.initHidden()

    for i in range(line_tensor.size()[0]):
        output, hidden = rnn(line_tensor[i], hidden)
        
      

    return output

for x1,y1 in test_data:
    
    label_tensor = torch.tensor([y1],dtype =torch.long)
    review_tensor = mymodel_reviewtensor(x1)
    output = evaluate(review_tensor)
    
    label_pred = output_predict_y(output)
    confusion[y1][label_pred] += 1


In [224]:
accuracy_my_model_binary = confusion.trace()/confusion.sum()

In [225]:
print("the accuracy for my model with RNN binary is ",accuracy_my_model_binary)

the accuracy for my model with RNN binary is  tensor(0.8205)


## binary RNN with google 300 model 

In [227]:
import gensim.downloader as api

opm = api.load("word2vec-google-news-300")

In [228]:
def review2tensor(x):
    max_review = 50
    ### create a matrix 
    review = torch.zeros(max_review,1,300)
    pad = torch.zeros(300)
    index = 0
    for i in x:
        if i in opm and index < 50:
            vector = opm [i]
            review[index][0] = torch.from_numpy(vector)
            index += 1
        else:
            continue
    if index < 50:
        review[index][0] = pad
    review = torch.nan_to_num(review)
    return review 
        

In [229]:
review2tensor('the').size()[0]

50

In [230]:
review2tensor(binary['review_body'].iloc[49])[10]

tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
         0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.

In [231]:
learning_rate = 0.005 # If you set this too high, it might explode. If too low, it might not learn

def train(category_tensor, line_tensor):
    hidden = rnn.initHidden()

    rnn.zero_grad()

    for i in range(line_tensor.size()[0]):
        output, hidden = rnn(line_tensor[i], hidden)

    loss = criterion(output, category_tensor)
    loss.backward()
    torch.nn.utils.clip_grad_norm_(rnn.parameters(),0.5)
    # Add parameters' gradients to their values, multiplied by learning rate
    for p in rnn.parameters():
        p.data.add_(p.grad.data, alpha=-learning_rate)

    return output, loss.item()

In [232]:
X = binary['review_body']
Y = binary['y']

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 19, test_size = 0.2)

x_train = np.array(x_train.tolist())

y_train = np.array(y_train.tolist())

x_train = np.nan_to_num(x_train)

x_test = np.array(x_test.tolist())

y_test = np.array(y_test.tolist())

x_test = np.nan_to_num(x_test)

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:8: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:14: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  


In [233]:
from torch.utils.data import DataLoader, Dataset
from torch.utils.data import TensorDataset, DataLoader

In [234]:
import torch.nn as nn

class RNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(RNN, self).__init__()

        self.hidden_size = hidden_size

        self.i2h = nn.Linear(input_size + hidden_size, hidden_size)
        self.i2o = nn.Linear(input_size + hidden_size, output_size)
        self.softmax = nn.LogSoftmax(dim=1)

    def forward(self, input, hidden):
        combined = torch.cat((input, hidden), 1)
        hidden = self.i2h(combined)
        output = self.i2o(combined)
        output = self.softmax(output)
        return output, hidden

    def initHidden(self):
        return torch.zeros(1, self.hidden_size)

rnn = RNN(300, 50, 2)

In [235]:
input = review2tensor(binary['review_body'].iloc[4])
hidden = torch.zeros(1, 50)

In [236]:
output, next_hidden = rnn(input[0], hidden)

In [237]:
def output_predict_y(output):
    top_n, top_i = output.topk(1)
    y = top_i[0].item()
    return y

In [238]:
yfromdata(output)

0

In [239]:
criterion = nn.NLLLoss()

In [240]:
learning_rate = 0.005 # If you set this too high, it might explode. If too low, it might not learn

def train(category_tensor, line_tensor):
    hidden = rnn.initHidden()

    rnn.zero_grad()

    for i in range(line_tensor.size()[0]):
        output, hidden = rnn(line_tensor[i], hidden)

    loss = criterion(output, category_tensor)
    loss.backward()
    torch.nn.utils.clip_grad_norm_(rnn.parameters(),0.5)
    # Add parameters' gradients to their values, multiplied by learning rate
    for p in rnn.parameters():
        p.data.add_(p.grad.data, alpha=-learning_rate)

    return output, loss.item()

In [241]:
from torch.utils.data import DataLoader, Dataset
from torch.utils.data import TensorDataset, DataLoader

In [242]:
main_data = list(zip(x_train,y_train))

In [243]:
# Keep track of losses for plotting
current_loss = 0

def timeSince(since):
    now = time.time()
    s = now - since
    m = math.floor(s / 60)
    s -= m * 60
    return '%dm %ds' % (m, s)

start = time.time()
ii=0
correct = 0
for x_i,y_i in main_data:
    line_tensor = review2tensor(x_i)
    y_i = torch.tensor([y_i],dtype =torch.long)
    output, loss = train(y_i, line_tensor)
    current_loss += loss
    ii += 1
    predict_y = yfromdata(output)
    if y_i == predict_y:
        correct +=1
    if ii %2500 == 0 :
        print(ii,"current avg lose is ",current_loss/ii,'time is ',time.time() - start)
        

2500 current avg lose is  0.6946149057388306 time is  22.118892192840576
5000 current avg lose is  0.6957750918149949 time is  39.81695795059204
7500 current avg lose is  0.6966212011496226 time is  56.48860502243042
10000 current avg lose is  0.69695377741158 time is  73.53538584709167
12500 current avg lose is  0.7033569383120537 time is  93.37862300872803
15000 current avg lose is  0.7123965916832288 time is  114.35654783248901
17500 current avg lose is  0.72068262171575 time is  131.81934809684753
20000 current avg lose is  0.7233350892782211 time is  148.8105869293213
22500 current avg lose is  0.7266608888400925 time is  166.8097460269928
25000 current avg lose is  0.7288242967271805 time is  184.70361185073853
27500 current avg lose is  0.7300440458850427 time is  203.6105260848999
30000 current avg lose is  0.731794490335385 time is  221.02341198921204
32500 current avg lose is  0.7336278566360473 time is  238.65603375434875
35000 current avg lose is  0.7356945490990366 time is

In [244]:
loss

0.0005920564290136099

In [245]:
current_loss

208765.9773401733

In [246]:
test_data = list(zip(x_test,y_test))

In [247]:

n_label = 2
confusion = torch.zeros(n_label, n_label)

def evaluate(line_tensor):
    hidden = rnn.initHidden()

    for i in range(line_tensor.size()[0]):
        output, hidden = rnn(line_tensor[i], hidden)
        
      

    return output

for x1,y1 in test_data:
    
    label_tensor = torch.tensor([y1],dtype =torch.long)
    review_tensor = review2tensor(x1)
    output = evaluate(review_tensor)
    
    label_pred = output_predict_y(output)
    confusion[y1][label_pred] += 1


In [248]:
accuracy_googel_300_binary = confusion.trace()/confusion.sum()

In [249]:
print("the accuracy for googel 300 model with RNN binary is ",accuracy_googel_300_binary)

the accuracy for googel 300 model with RNN binary is  tensor(0.7890)


## trinary with my model RNN

In [250]:
def mymodel_reviewtensor(x):
    max_review = 50
    ### create a matrix 
    review = torch.zeros(max_review,1,300)
    pad = torch.zeros(300)
    index = 0
    for i in x:
        if i in model_1.wv and index < 50:
            vector = model_1.wv [i]
            review[index][0] = torch.from_numpy(vector)
            index += 1
        else:
            continue
    if index < 50:
        review[index][0] = pad
    review = torch.nan_to_num(review)
    return review 

In [252]:
X = new_test['review_body']
Y = new_test['y']

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 19, test_size = 0.2)

x_train = np.array(x_train.tolist())

y_train = np.array(y_train.tolist())

x_train = np.nan_to_num(x_train)

x_test = np.array(x_test.tolist())

y_test = np.array(y_test.tolist())

x_test = np.nan_to_num(x_test)

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:8: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:14: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  


In [254]:
import torch.nn as nn

class RNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(RNN, self).__init__()

        self.hidden_size = hidden_size

        self.i2h = nn.Linear(input_size + hidden_size, hidden_size)
        self.i2o = nn.Linear(input_size + hidden_size, output_size)
        self.softmax = nn.LogSoftmax(dim=1)

    def forward(self, input, hidden):
        combined = torch.cat((input, hidden), 1)
        hidden = self.i2h(combined)
        output = self.i2o(combined)
        output = self.softmax(output)
        return output, hidden

    def initHidden(self):
        return torch.zeros(1, self.hidden_size)

rnn = RNN(300, 50, 3)

In [255]:
input = review2tensor(binary['review_body'].iloc[4])
hidden = torch.zeros(1, 50)

In [256]:
criterion = nn.NLLLoss()

In [257]:
learning_rate = 0.005 # If you set this too high, it might explode. If too low, it might not learn

def train(category_tensor, line_tensor):
    hidden = rnn.initHidden()

    rnn.zero_grad()

    for i in range(line_tensor.size()[0]):
        output, hidden = rnn(line_tensor[i], hidden)

    loss = criterion(output, category_tensor)
    loss.backward()
    torch.nn.utils.clip_grad_norm_(rnn.parameters(),0.5)
    # Add parameters' gradients to their values, multiplied by learning rate
    for p in rnn.parameters():
        p.data.add_(p.grad.data, alpha=-learning_rate)

    return output, loss.item()

In [258]:
main_data = list(zip(x_train,y_train))

In [259]:
# Keep track of losses for plotting
current_loss = 0

def timeSince(since):
    now = time.time()
    s = now - since
    m = math.floor(s / 60)
    s -= m * 60
    return '%dm %ds' % (m, s)

start = time.time()
ii=0
correct = 0
for x_i,y_i in main_data:
    line_tensor = review2tensor(x_i)
    y_i = torch.tensor([y_i],dtype =torch.long)
    output, loss = train(y_i, line_tensor)
    current_loss += loss
    ii += 1
    if ii %2500 == 0 :
        print(ii,"current avg lose is ",current_loss/ii,'time is ',time.time() - start)

2500 current avg lose is  1.058883339238167 time is  32.93949484825134
5000 current avg lose is  1.0648978867650032 time is  69.53135681152344
7500 current avg lose is  1.0671727425018946 time is  101.4669828414917
10000 current avg lose is  1.070502868539095 time is  137.4072709083557
12500 current avg lose is  1.0674601861286164 time is  171.49279308319092
15000 current avg lose is  1.0676490481615066 time is  212.4065079689026
17500 current avg lose is  1.0673559479747499 time is  238.32458591461182
20000 current avg lose is  1.0682490519195795 time is  265.84134697914124
22500 current avg lose is  1.0685526640786065 time is  296.73723793029785
25000 current avg lose is  1.067924862034321 time is  340.1728708744049
27500 current avg lose is  1.0683107580510052 time is  377.68309903144836
30000 current avg lose is  1.0681639694203933 time is  402.9458739757538
32500 current avg lose is  1.06844231048639 time is  435.8811500072479
35000 current avg lose is  1.0687974279531411 time is 

In [260]:
test_data = list(zip(x_test,y_test))

In [261]:

n_label = 3
confusion = torch.zeros(n_label, n_label)

def evaluate(line_tensor):
    hidden = rnn.initHidden()

    for i in range(line_tensor.size()[0]):
        output, hidden = rnn(line_tensor[i], hidden)
        
      

    return output

for x1,y1 in test_data:
    
    label_tensor = torch.tensor([y1],dtype =torch.long)
    review_tensor = review2tensor(x1)
    output = evaluate(review_tensor)
    
    label_pred = output_predict_y(output)
    confusion[y1][label_pred] += 1


In [262]:
accuracy_my_model_ternary = confusion.trace()/confusion.sum()

In [263]:
accuracy_my_model_ternary

tensor(0.6221)

In [264]:
print("the accuracy for my model with RNN binary is ",accuracy_my_model_ternary)

the accuracy for my model with RNN binary is  tensor(0.6221)


## trinary with google 300 model RNN

In [268]:
def opm_reviewtensor(x):
    max_review = 50
    ### create a matrix 
    review = torch.zeros(max_review,1,300)
    pad = torch.zeros(300)
    index = 0
    for i in x:
        if i in opm and index < 50:
            vector = opm [i]
            review[index][0] = torch.from_numpy(vector)
            index += 1
        else:
            continue
    if index < 50:
        review[index][0] = pad
    review = torch.nan_to_num(review)
    return review 

In [269]:
import torch.nn as nn

class RNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(RNN, self).__init__()

        self.hidden_size = hidden_size

        self.i2h = nn.Linear(input_size + hidden_size, hidden_size)
        self.i2o = nn.Linear(input_size + hidden_size, output_size)
        self.softmax = nn.LogSoftmax(dim=1)

    def forward(self, input, hidden):
        combined = torch.cat((input, hidden), 1)
        hidden = self.i2h(combined)
        output = self.i2o(combined)
        output = self.softmax(output)
        return output, hidden

    def initHidden(self):
        return torch.zeros(1, self.hidden_size)

rnn = RNN(300, 50, 3)

In [270]:
criterion = nn.NLLLoss()

In [271]:
learning_rate = 0.005 # If you set this too high, it might explode. If too low, it might not learn

def train(category_tensor, line_tensor):
    hidden = rnn.initHidden()

    rnn.zero_grad()

    for i in range(line_tensor.size()[0]):
        output, hidden = rnn(line_tensor[i], hidden)

    loss = criterion(output, category_tensor)
    loss.backward()
    torch.nn.utils.clip_grad_norm_(rnn.parameters(),0.5)
    # Add parameters' gradients to their values, multiplied by learning rate
    for p in rnn.parameters():
        p.data.add_(p.grad.data, alpha=-learning_rate)

    return output, loss.item()

In [272]:
main_data = list(zip(x_train,y_train))

In [273]:
# Keep track of losses for plotting
current_loss = 0

def timeSince(since):
    now = time.time()
    s = now - since
    m = math.floor(s / 60)
    s -= m * 60
    return '%dm %ds' % (m, s)

start = time.time()
ii=0
correct = 0
for x_i,y_i in main_data:
    line_tensor = opm_reviewtensor(x_i)
    y_i = torch.tensor([y_i],dtype =torch.long)
    output, loss = train(y_i, line_tensor)
    current_loss += loss
    ii += 1
    if ii %2500 == 0 :
        print(ii,"current avg lose is ",current_loss/ii,'time is ',time.time() - start)

2500 current avg lose is  1.0594981743097305 time is  23.336733102798462
5000 current avg lose is  1.064748005270958 time is  46.70505118370056
7500 current avg lose is  1.0665473565498989 time is  63.84702515602112
10000 current avg lose is  1.069900180619955 time is  81.04995703697205
12500 current avg lose is  1.0668171494627 time is  98.48082613945007
15000 current avg lose is  1.0672467289964358 time is  116.14186596870422
17500 current avg lose is  1.066933060513224 time is  133.63144707679749
20000 current avg lose is  1.067680196505785 time is  150.6814329624176
22500 current avg lose is  1.0679751199086507 time is  167.5362868309021
25000 current avg lose is  1.067366377673149 time is  184.75301909446716
27500 current avg lose is  1.067827658234943 time is  201.7656970024109
30000 current avg lose is  1.0678597442567348 time is  218.99542498588562
32500 current avg lose is  1.0681703750610352 time is  236.04149508476257
35000 current avg lose is  1.0684610444409506 time is  25

In [274]:
accuracy_opm_ternary = confusion.trace()/confusion.sum()

In [275]:
print("the accuracy for google model with RNN binary is ",accuracy_opm_ternary)

the accuracy for google model with RNN binary is  tensor(0.6221)


#RNN

RNN : Binary Case with google model | The Test Accuracy is 0.7890

RNN : Ternary case with googel model | the Test Accuracy is 0.6221

RNN: Binary case with my own model | Test Accuracy is 0.8205

RNN: Ternary case with my own model | Test Accuracy is 0.6223

